### PR TITLE
HSHARDS-66 fix org.hibernate.TransactionException in tests

### PR DESCRIPTION
**Pull request by alex-dukhno**

### CODE REVIEW

**Comment by alex-dukhno** [comment]
Need to redevelop transytion after failed commit should be rolled back.

@brmeyer or after all failed commit transaction could be committed? Because 
this condition
    begun && !(rolledBack || committed);
conflict with 
    if ( commitFailed ) {
        throw new TransactionException( "cannot re-start transaction after failed commit" );
    }
in ShardedTransactionImpl.begin()

**Comment by alex-dukhno** [comment]
Problem was fixed, just changed sequence of checking transaction status

File: src/main/java/org/hibernate/shards/transaction/ShardedTransactionImpl.java [hunk 1]
```diff
@@ -148,6 +149,20 @@ public void commit() throws HibernateException {
 		committed = true;
 	}
 
+	private void afterTransactionCompletion(final int status) {
+		begun = Status.STATUS_COMMITTED != status;
```
Comment (brmeyer) [inline src/main/java/org/hibernate/shards/transaction/ShardedTransactionImpl.java]: A little confused here.  Originally, we simply set begun = false.  If a commit exception is caught, afterTransactionCompletion will be given Status.STATUS_UNKNOWN.  In that type of case, it probably doesn't make sense to let begun == true.

Comment (alex-dukhno) [inline src/main/java/org/hibernate/shards/transaction/ShardedTransactionImpl.java]: AFAIR method **afterTransactionCompletion** is invoked also after fail on committing transaction and if **begin** is set to **false** in method **session.getTransaction** we get new transaction without roll backed previous one. If **begin** is set to **true** we get **org.hibernate.TransactionException: nested transactions not supported**. It should be deeper investigated. Changes affect many tests and implementations. I will add _TODO_ and create ticket on JIRA

**Comment by alex-dukhno** [comment]
@brmeyer  what do you think about this implementation of ShardedTransaction?

File: src/main/java/org/hibernate/shards/transaction/ShardedTransactionImpl.java [hunk 2]
```diff
@@ -107,156 +88,145 @@ public void begin() throws HibernateException {
...
 
 	@Override
-	public boolean isParticipating() {
+	public void markRollbackOnly() {
+
+	}
+
+	@Override
+	public boolean isActive() {
+		return getLocalStatus() == LocalStatus.FAILED_COMMIT
```
Comment (brmeyer) [inline src/main/java/org/hibernate/shards/transaction/ShardedTransactionImpl.java:179]: Hey @Alexander-Dukhno, why are you including FAILED_COMMIT here?

**Comment by brmeyer** [comment]
@Alexander-Dukhno, this is a gigantic improvement -- great work.  My only question is above, regarding FAILED_COMMIT.  Thanks!

File: src/main/java/org/hibernate/shards/transaction/ShardedTransactionImpl.java (hunk 2 above)
Comment (alex-dukhno) [inline src/main/java/org/hibernate/shards/transaction/ShardedTransactionImpl.java:179]: I don't know why :)
Previous developers decided that transaction after failed commit should be active and you can only roll back it. I haven't enough experience to solve this issue better

**Pull request closed without merging**